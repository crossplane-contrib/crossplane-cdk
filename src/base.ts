import * as cdk8s from 'cdk8s';
import { Construct } from 'constructs';

/**
 * Initialization properties for resources.
 */
export interface ResourceProps {
  /**
   * @schema ResourceProps#name
   */
  readonly name?: string;

  /**
   * Metadata that all persisted resources must have, which includes all objects
   * users must create.
   * @schema ResourceProps#metadata
   */
  readonly metadata?: cdk8s.ApiObjectMetadata;
}

/**
 * Represents a resource.
 */
export interface IResource {
  /**
   * The Kubernetes name of this resource.
   */
  readonly name: string;
}

/**
 * Base class for all Crossplane core objects. Represents a single resource.
 */
export abstract class Resource extends Construct implements IResource {

  /**
   * The underlying cdk8s API object.
   */
  protected abstract readonly apiObject: cdk8s.ApiObject;

  constructor(scope: Construct, id: string, _: ResourceProps) {
    super(scope, id);
  }

  /**
   * The name of this API object.
   */
  public get name(): string {
    return this.apiObject.name;
  }
}