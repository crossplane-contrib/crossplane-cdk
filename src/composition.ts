import * as cdk8s from 'cdk8s';
import { Construct } from 'constructs';
import { Resource, ResourceProps } from './base';
import * as l1 from './imports/apiextensions.crossplane.io';
import { undefinedIfEmpty } from './utils';
import { CompositeResourceDefinition } from './xrd';

/**
 * Composition defines the group of resources to be created when a compatible type is created with reference to the composition.
 */
export class Composition extends Resource {
  /**
   * Defines a "CompositeResourceDefinition" API object
   * @param scope the scope in which to define this object
   * @param id a scope-local name for the object
   * @param props initialiation props - note: currently only .name and .metadata are processed
   */

  /**
   * @see base.Resource.apiObject
   */
  protected readonly apiObject: cdk8s.ApiObject;

  private _xrd: CompositeResourceDefinition;
  private _resources: CompositionSpecResource[] = [];
  private _writeConnectionSecretToNamespace: string;

  public constructor(scope: Construct, id: string, xrd: CompositeResourceDefinition, props: CompositionProps = {}) {
    super(scope, id, { metadata: props.metadata });

    this._xrd = xrd;
    this._writeConnectionSecretToNamespace = xrd.meta.connectionSecret?.defaultNamespace ?? 'crossplane-system';

    const name = props.name ?? (props.metadata?.name ?? id);

    let annotations: any = {};
    if (props.metadata?.annotations) { annotations = { ...props.metadata.annotations, ...annotations }; }

    //TODO: load prop.spec for those not using fluent API

    this.apiObject = new l1.Composition(this, name, {
      metadata: {
        ...props.metadata,
        name: name,
        annotations: annotations,
      },
      spec: cdk8s.Lazy.any({ produce: () => this._toL1() }),
    });
  }

  private _toL1(): l1.CompositionSpec {
    const xrd = this._xrd.meta;

    const resources = this._resourcesToL1();
    if (resources === undefined) {
      throw new Error(`at least one resource is required for '${this.name}'`);
    }

    return {
      writeConnectionSecretsToNamespace: this._writeConnectionSecretToNamespace,
      compositeTypeRef: {
        apiVersion: `${xrd.group}/${xrd.versionServed}`,
        kind: xrd.names.kind,
      },
      resources: resources,
    };
  }

  private _resourcesToL1(): l1.CompositionSpecResources[] | undefined {
    const result = new Array();
    for (const resource of this._resources) {
      result.push(resource._toL1());
    }
    return undefinedIfEmpty(result);
  }

  public addResource(base: any): CompositionSpecResource {
    const resource = new CompositionSpecResource(this._xrd, base);
    this._resources.push(resource);
    return resource;
  }
}

export class CompositionSpecResource {
  private _xrd: CompositeResourceDefinition;
  private _base: any;
  private _patches: l1.CompositionSpecResourcesPatches[] = [];
  private _connectionDetails: l1.CompositionSpecResourcesConnectionDetails[] = [];

  /**
   * use Composition.addResource() instead
   * @param xrd
   *
   * @internal
   */
  public constructor(xrd: CompositeResourceDefinition, base: any) {
    this._xrd = xrd;
    this._base = base;
  }

  /**
   * @internal
   */
  public _toL1(): l1.CompositionSpecResources {
    return {
      base: this._base,
      patches: undefinedIfEmpty(this._patches),
      connectionDetails: undefinedIfEmpty(this._connectionDetails),
    };
  }

  public mapFieldPath(from: string, to: string): CompositionSpecResource {
    this._patches.push({
      fromFieldPath: from,
      toFieldPath: to,
    });
    return this;
  }

  /**
   *
   * @param from
   * @param format e.g. "%s-suffix"
   * @param to
   */
  public mapFieldPathXFormStringFormat(from: string, format: string, to: string): CompositionSpecResource {
    this._patches.push({
      fromFieldPath: from,
      toFieldPath: to,
      transforms: [
        {
          type: 'string',
          string: {
            fmt: format,
          },
        },
      ],
    });
    return this;
  }

  /*
  public connectionDetails(details: l1.CompositionSpecResourcesConnectionDetails): CompositionSpecResource {
    this._connectionDetails.push(details);
    return this;
  }
  */

  public connectionDetailsFromXrd(xrd: CompositeResourceDefinition = this._xrd): CompositionSpecResource {
    const keys = xrd.meta.connectionSecret?.keys ?? [];
    if (keys.length == 0) {
      throw new Error(`no connectionSecret.keys found for '${xrd.fqn}'`);
    }

    for (const k of keys) {
      this._connectionDetails.push({ fromConnectionSecretKey: k });
    }

    return this;
  }
}

/**
 * Composition Props
 */

/**
 * Composition defines the group of resources to be created when a compatible type is created with reference to the composition.
 *
 * TODO: Support all available props in fluent API.
 * TODO: Process props.spec in Composition so you can init with
 * static spec fields instead of using the fluent API.
 */
export interface CompositionProps extends ResourceProps {
  /**
   * CompositionSpecProps specifies the desired state of the definition.
   *
   * @schema CompositionProps#spec
   */
  //readonly spec?: CompositionSpecProps;

}