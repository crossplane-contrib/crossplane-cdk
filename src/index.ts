export * from './base';
export * from './configuration';
export * from './xrd';