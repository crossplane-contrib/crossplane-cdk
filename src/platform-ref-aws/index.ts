import { App, Chart } from 'cdk8s';
import { Configuration } from '../crossplane/imports/meta.pkg.crossplane.io';
import databasePostgresDefinition from './database/postgres/definition'
import databasePostgresComposition from './database/postgres/composition'
import networkDefinition from './network/definition'
import networkComposition from './network/composition'
import clusterDefinition from './cluster/definition'

const crossplanePackage = new App();

const crossplaneYaml = new Chart(crossplanePackage, 'crossplane');
new Configuration(crossplaneYaml, 'platform-ref-aws', { 
  metadata: {
    name: "platform-ref-aws",
    annotations: {
      company: "Upbound",
      maintainer: "Jared Watts <jared@upbound.io>",
      keywords: "aws, cloud-native, kubernetes, example, platform, reference",
      source: "github.com/upbound/platform-ref-aws",
      license: "Apache-2.0",
      descriptionShort: "The AWS reference platform for Kubernetes and Data Services.\n",
      description: "This reference platform Configuration for Kubernetes and Data Services\nis a starting point to build, run, and operate your own internal cloud\nplatform and offer a self-service console and API to your internal teams.\n",
      readme: "This reference platform `Configuration` for Kubernetes and Data Services\nis a starting point to build, run, and operate your own internal cloud\nplatform and offer a self-service console and API to your internal teams.\nIt provides platform APIs to provision fully configured EKS clusters,\nwith secure networking, and stateful cloud services (RDS) designed to\nsecurely connect to the nodes in each EKS cluster -- all composed using\ncloud service primitives from the [Crossplane AWS\nProvider](https://doc.crds.dev/github.com/crossplane/provider-aws). App\ndeployments can securely connect to the infrastructure they need using\nsecrets distributed directly to the app namespace.\n\n[Quickstart\nGuide](https://github.com/upbound/platform-ref-aws/#quick-start)\n\n[APIs in this\nConfiguration](https://github.com/upbound/platform-ref-aws/#apis-in-this-configuration)\n\nTo learn more checkout the [GitHub\nrepo](https://github.com/upbound/platform-ref-aws/) that you can copy and\ncustomize to meet the exact needs of your organization!\n"
    }
  },
  spec: {
    crossplane: {
      version: ">=v0.14.0-0"
    },
    dependsOn: [
      {
        provider: "crossplane/provider-aws",
        version: ">=v0.15.0-0"
      },
      {
        provider: "crossplane/provider-helm",
        version: ">=v0.3.7-0"
      }
    ]
  }
});

databasePostgresDefinition(crossplanePackage);
databasePostgresComposition(crossplanePackage);

networkDefinition(crossplanePackage);
networkComposition(crossplanePackage);

clusterDefinition(crossplanePackage);

crossplanePackage.synth();