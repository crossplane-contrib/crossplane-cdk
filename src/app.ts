import { App, Chart } from 'cdk8s';
import { Configuration } from '.';
import { CompositeResourceDefinition } from './apiextensions.crossplane.io/compositeresourcedefinition';
import { Composition } from './apiextensions.crossplane.io/composition';

const app = new App();
const chart = new Chart(app, 'test');

const config = new Configuration(chart, 'config', {
  company: 'Upbound',
  keywords: ['aws', 'cloud-native', 'kubernetes', 'example', 'platform', 'reference'],
});

config.addProvider('crossplane/provider-aws', '>=v0.14.0-0');

new CompositeResourceDefinition(chart, 'xrd', {
  metadata: {
    name: 'compositepostgresqlinstances.aws.platformref.crossplane.io',
  },
  spec: {
    claimNames: {
      kind: 'PostgreSQLInstance',
      plural: 'postgresqlinstances',
    },
    connectionSecretKeys: [
      'username',
      'password',
      'endpoint',
      'port',
    ],
    group: 'aws.platformref.crossplane.io',
    names: {
      kind: 'CompositePostgreSQLInstance',
      plural: 'compositepostgresqlinstances',
    },
    versions: [
      {
        name: 'v1alpha1',
        served: true,
        referenceable: true,
        schema: {
          openAPIV3Schema: {
            type: 'object',
            properties: {
              spec: {
                type: 'object',
                properties: {
                  parameters: {
                    type: 'object',
                    properties: {
                      storageGB: { type: 'integer' },
                      networkRef: {
                        type: 'object',
                        description: 'A reference to the Network object that this postgres should be connected to.',
                        properties: {
                          id: { type: 'string', description: 'ID of the Network object this ref points to.' },
                        },
                        required: ['id'],
                      },
                    },
                    required: [
                      'storageGB',
                      'networkRef',
                    ],
                  },
                },
                required: ['parameters'],
              },
            },
          },
        },
      },
    ],
  },
});

new Composition(chart, 'my-composition', {
  metadata: {
    name: 'compositepostgresqlinstances.aws.platformref.crossplane.io',
    labels: {
      provider: 'aws',
    },
  },
  spec: {
    writeConnectionSecretsToNamespace: 'crossplane-system',
    compositeTypeRef: {
      apiVersion: 'aws.platformref.crossplane.io/v1alpha1',
      kind: 'CompositePostgreSQLInstance',
    },
    resources: [
      {
        base: {
          apiVersion: 'database.aws.crossplane.io/v1beta1',
          kind: 'DBSubnetGroup',
          spec: {
            forProvider: {
              region: 'us-west-2',
              description: 'An excellent formation of subnetworks.',
            },
            reclaimPolicy: 'Delete',
          },
        },
        patches: [
          {
            fromFieldPath: 'spec.parameters.networkRef.id',
            toFieldPath: 'spec.forProvider.subnetIdSelector.matchLabels[networks.aws.platformref.crossplane.io/network-id]',
          },
        ],
      },
      {
        base: {
          apiVersion: 'database.aws.crossplane.io/v1beta1',
          kind: 'RDSInstance',
          spec: {
            forProvider: {
              region: 'us-west-2',
              dbSubnetGroupNameSelector: {
                matchControllerRef: true,
              },
              dbInstanceClass: 'db.t2.small',
              masterUsername: 'masteruser',
              engine: 'postgres',
              engineVersion: '9.6',
              skipFinalSnapshotBeforeDeletion: true,
              publiclyAccessible: false,
            },
            writeConnectionSecretToRef: {
              namespace: 'crossplane-system',
            },
            reclaimPolicy: 'Delete',
          },
        },
        patches: [
          {
            fromFieldPath: 'metadata.uid',
            toFieldPath: 'spec.writeConnectionSecretToRef.name',
            transforms: [
              {
                type: 'string',
                string: {
                  fmt: '%s-postgresql',
                },
              },
            ],
          },
          {
            fromFieldPath: 'spec.parameters.storageGB',
            toFieldPath: 'spec.forProvider.allocatedStorage',
          },
          {
            fromFieldPath: 'spec.parameters.networkRef.id',
            toFieldPath: 'spec.forProvider.vpcSecurityGroupIDSelector.matchLabels[networks.aws.platformref.crossplane.io/network-id]',
          },
        ],
        connectionDetails: [
          {
            fromConnectionSecretKey: 'username',
          },
          {
            fromConnectionSecretKey: 'password',
          },
          {
            fromConnectionSecretKey: 'endpoint',
          },
          {
            fromConnectionSecretKey: 'port',
          },
        ],
      },
    ],
  },
});

app.synth();