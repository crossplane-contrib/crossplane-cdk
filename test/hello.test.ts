import { Testing } from 'cdk8s';
import { Configuration } from '../src';
import { CompositeResourceDefinition } from '../src/apiextensions.crossplane.io/compositeresourcedefinition';

test('hello', () => {
  const chart = Testing.chart();

  const config = new Configuration(chart, 'my-config', {
    company: 'Upbound',
    keywords: ['aws', 'cloud-native', 'kubernetes', 'example', 'platform', 'reference'],
  });

  config.addProvider('crossplane/provider-aws', '>=v0.14.0-0');

  expect(Testing.synth(chart)).toMatchSnapshot();
});

test('xrd', () => {
  const chart = Testing.chart();

  new CompositeResourceDefinition(chart, 'xrd', {
    metadata: {
      name: 'compositepostgresqlinstances.aws.platformref.crossplane.io',
    },
    spec: {
      claimNames: {
        kind: 'PostgreSQLInstance',
        plural: 'postgresqlinstances',
      },
      connectionSecretKeys: [
        'username',
        'password',
        'endpoint',
        'port',
      ],
      group: 'aws.platformref.crossplane.io',
      names: {
        kind: 'CompositePostgreSQLInstance',
        plural: 'compositepostgresqlinstances',
      },
      versions: [
        {
          name: 'v1alpha1',
          served: true,
          referenceable: true,
          schema: {
            openAPIV3Schema: {
              type: 'object',
              properties: {
                spec: {
                  type: 'object',
                  properties: {
                    parameters: {
                      type: 'object',
                      properties: {
                        storageGB: { type: 'integer' },
                        networkRef: {
                          type: 'object',
                          description: 'A reference to the Network object that this postgres should be connected to.',
                          properties: {
                            id: { type: 'string', description: 'ID of the Network object this ref points to.' },
                          },
                          required: ['id'],
                        },
                      },
                      required: [
                        'storageGB',
                        'networkRef',
                      ],
                    },
                  },
                  required: ['parameters'],
                },
              },
            },
          },
        },
      ],
    },
  });

  expect(Testing.synth(chart)).toMatchSnapshot();
});
