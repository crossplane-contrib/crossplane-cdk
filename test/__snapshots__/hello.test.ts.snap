// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hello 1`] = `
Array [
  Object {
    "apiVersion": "meta.pkg.crossplane.io/v1alpha1",
    "kind": "Configuration",
    "metadata": Object {
      "annotations": Object {
        "company": "Upbound",
        "keywords": "aws, cloud-native, kubernetes, example, platform, reference",
      },
      "name": "my-config",
    },
    "spec": Object {
      "crossplane": Object {
        "version": ">=v0.14.0-0",
      },
      "dependsOn": Array [
        Object {
          "provider": "crossplane/provider-aws",
          "version": ">=v0.14.0-0",
        },
      ],
    },
  },
]
`;

exports[`xrd 1`] = `
Array [
  Object {
    "apiVersion": "apiextensions.crossplane.io/v1alpha1",
    "kind": "CompositeResourceDefinition",
    "metadata": Object {
      "name": "compositepostgresqlinstances.aws.platformref.crossplane.io",
    },
    "spec": Object {
      "claimNames": Object {
        "kind": "PostgreSQLInstance",
        "plural": "postgresqlinstances",
      },
      "connectionSecretKeys": Array [
        "username",
        "password",
        "endpoint",
        "port",
      ],
      "group": "aws.platformref.crossplane.io",
      "names": Object {
        "kind": "CompositePostgreSQLInstance",
        "plural": "compositepostgresqlinstances",
      },
      "versions": Array [
        Object {
          "name": "v1alpha1",
          "referenceable": true,
          "schema": Object {
            "openAPIV3Schema": Object {
              "properties": Object {
                "spec": Object {
                  "properties": Object {
                    "parameters": Object {
                      "properties": Object {
                        "networkRef": Object {
                          "description": "A reference to the Network object that this postgres should be connected to.",
                          "properties": Object {
                            "id": Object {
                              "description": "ID of the Network object this ref points to.",
                              "type": "string",
                            },
                          },
                          "required": Array [
                            "id",
                          ],
                          "type": "object",
                        },
                        "storageGB": Object {
                          "type": "integer",
                        },
                      },
                      "required": Array [
                        "storageGB",
                        "networkRef",
                      ],
                      "type": "object",
                    },
                  },
                  "required": Array [
                    "parameters",
                  ],
                  "type": "object",
                },
              },
              "type": "object",
            },
          },
          "served": true,
        },
      ],
    },
  },
]
`;
